### PR TITLE
Self influence

### PR DESCRIPTION
**Pull request by tfjgeorge**

### CODE REVIEW

File: mislabeled/probe/_linear.py [hunk 1]
```diff
@@ -46,6 +47,99 @@
...
+        # gradients w.r.t. the parameters (weight, intercept)
+        dl_dy = self.grad_y(X, y)
+
+        X_p = X
+        if self.intercept is not None:
+            X_p = np.hstack((X, np.ones((X.shape[0], 1))))
+
+        return dl_dy[:, :, None] * X_p[:, None, :]
+
+    def hessian(self, X, y):
```
Comment (pierrenodet) [inline mislabeled/probe/_linear.py:100]: does it scale to models used in the benchmark? the previous sparse implementation was meant to do that.

File: mislabeled/probe/_linear.py [hunk 2]
```diff
@@ -135,10 +267,12 @@ def linearize_trees(
...
     if is_classifier(estimator):
         linear = default_linear_model["classification"]
+        linear.fit(leaves, y)
+        return linearize_linear_model_logreg(linear, leaves, y)
     else:
         linear = default_linear_model["regression"]
-    linear.fit(leaves, y)
-    return linearize_linear_model(linear, leaves, y)
+        linear.fit(leaves, y)
+        return linearize_linear_model_ridge(linear, leaves, y)
```
Comment (pierrenodet) [inline mislabeled/probe/_linear.py]: you could call directly linearize(linear, leaves, y) and the dispatch will call the right method automatically.

**Comment by tfjgeorge** [comment]
in summary, this unifies the @linear api to a single LinearModel class, with 'log_loss' and 'l2' loss currently implemented:
 - adds tests vs LOO
 - adds support for binary classification (whereas previously it was handled as a multiclass classification using softmax)
 - grad_y and grad_p methods
 - the influence probe is renamed SelfInfluence
- removes untested classifiers
- updates TracIn and GradSimilarity in accordance

Possible regression: sparse feature matrices

File: mislabeled/probe/_influence.py [hunk 3]
```diff
@@ -135,47 +71,31 @@ def __call__(self, estimator, X, y):
...
 
 class L2Representer(Minimize):
     """L2 Representer values"""
 
     @linear
     def __call__(self, estimator, X, y):
-        grad_l2_loss = estimator.predict(X) - y
-        return np.abs(grad_l2_loss) * norm2(X)
+        grad = estimator.grad_y(X, y)
+        return np.abs(grad).sum(axis=1) * norm2(X)
```
Comment (pierrenodet) [inline mislabeled/probe/_influence.py]: would be nice if L2Representer could be merged with Representer as with other probes, but it's just a nitpick.

File: examples/plot_linearized.ipynb
Comment (pierrenodet) [inline examples/plot_linearized.ipynb:1]: It looks good (besides trees, but we can look into that later), but what's up with the scales in the last plot?

**Review by pierrenodet — COMMENTED** [review]
Overall looks very clean :)

File: mislabeled/probe/_linear.py [hunk 4]
```diff
@@ -82,37 +175,76 @@ def linearize_pipeline(estimator, X, y):
...
+    return linear, X, y
+
+
+# commented out while they are not used in tests
+# Lasso and Elasticnet are yet to be added
+# @linearize.register(ElasticNet)
+# @linearize.register(ElasticNetCV)
+# @linearize.register(Lasso)
+# @linearize.register(LassoCV)
+# @linearize.register(RidgeClassifierCV)
```
Comment (pierrenodet) [inline mislabeled/probe/_linear.py]: Should RidgeClassifierCV be uncommented ?

File: mislabeled/probe/_linear.py [hunk 5]
```diff
@@ -21,31 +22,123 @@
...
 from sklearn.neural_network import MLPClassifier, MLPRegressor
 from sklearn.neural_network._base import ACTIVATIONS
 from sklearn.pipeline import Pipeline
 from sklearn.preprocessing import OneHotEncoder
-from sklearn.svm import LinearSVC, LinearSVR
 from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
 from sklearn.utils import check_X_y
 
 
 class LinearModel(NamedTuple):
```
Comment (pierrenodet) [inline mislabeled/probe/_linear.py:42]: Should we add a grad_X in a following PR for sensitivity?

Comment (tfjgeorge) [reply mislabeled/probe/_linear.py:42]: https://github.com/Orange-OpenSource/mislabeled/issues/10

File: mislabeled/probe/_linear.py (hunk 4 above)
Comment (tfjgeorge) [reply mislabeled/probe/_linear.py]: I commented out all classifiers that were not used in the tests

File: examples/plot_linearized.ipynb
Comment (tfjgeorge) [reply examples/plot_linearized.ipynb:1]: no idea :-D I agree that it looks suspicious

File: mislabeled/probe/_linear.py (hunk 1 above)
Comment (pierrenodet) [reply mislabeled/probe/_linear.py:100]: Maybe create a new issue so that we can close this pull request.

Comment (pierrenodet) [reply mislabeled/probe/_linear.py:100]: #11

File: mislabeled/probe/_influence.py [hunk 6]
```diff
@@ -135,47 +72,24 @@ def __call__(self, estimator, X, y):
...
-
-
-class L2Representer(Minimize):
-    """L2 Representer values"""
+        grad = estimator.grad_y(X, y)
+        # grad observed
+        if estimator._is_binary():
+            grad_observed = np.abs(grad[:, 0])
+        else:
+            if is_classifier(estimator):
```
Comment (pierrenodet) [inline mislabeled/probe/_influence.py]: At this point, the estimator is a LinearModel, and I don't think is_classifier returns something meaningful for it.

Comment (tfjgeorge) [reply mislabeled/probe/_influence.py]: wow nice catch => replaced by loss check in next commit.

Ideally we would inherit "classifierness" from the parent classn though
